neuroquery/img_utils.py

```diff
@@ -63,25 +64,58 @@ def _coords_to_masked_map(coordinates, masker, fwhm, output, idx):
     output[idx] = masker.transform(img).squeeze()
 
 
+def _prepare_memmap(output_memmap_file, shape, context):
+    if output_memmap_file is None:
+        tmp_dir = context.enter_context(tempfile.TemporaryDirectory())
+        output_memmap_file = str(Path(tmp_dir).joinpath("memmap.dat"))
+        return_memmap = False
+    else:
+        output_memmap_file = str(output_memmap_file)
+        return_memmap = True
+    output = np.memmap(
+        output_memmap_file,
+        mode="w+",
+        dtype=np.float32,
+        shape=shape,
+    )
+    if not return_memmap:
+        context.enter_context(contextlib.closing(output._mmap))
+    return output, return_memmap
+
+
 def coordinates_to_maps(
-    coordinates, mask_img=None, target_affine=(4, 4, 4), fwhm=9.0, n_jobs=1
+    coordinates,
+    mask_img=None,
+    target_affine=(4, 4, 4),
+    fwhm=9.0,
+    n_jobs=1,
+    output_memmap_file=None,
 ):
     masker = get_masker(mask_img=mask_img, target_affine=target_affine)
     pmids = np.unique(coordinates["pmid"].values)
-    n_articles, n_voxels = len(pmids), image.get_data(masker.mask_img_).sum()
-    with tempfile.TemporaryDirectory() as tmp_dir:
-        tmp_file = Path(tmp_dir).joinpath("brain_maps_memmap.dat")
-        output = np.memmap(
-            tmp_file, mode="w+", dtype=np.float64, shape=(n_articles, n_voxels)
+    shape = len(pmids), image.get_data(masker.mask_img_).sum()
+    with contextlib.ExitStack() as context:
+        output, return_memmap = _prepare_memmap(
+            output_memmap_file, shape, context
         )
         all_articles = coordinates.groupby("pmid", sort=True)
         Parallel(n_jobs, verbose=1)(
             delayed(_coords_to_masked_map)(
-                article.loc[:, ["x", "y", "z"]].values, masker, fwhm, output, i
+                article.loc[:, ["x", "y", "z"]].values,
+                masker,
+                fwhm,
+                output,
+                i,
             )
             for i, (pmid, article) in enumerate(all_articles)
         )
-        return pd.DataFrame(np.array(output), index=pmids), masker
+        output.flush()
+        if return_memmap:
+            result = pd.DataFrame(output, index=pmids)
+        else:
+            result = pd.DataFrame(np.array(output), index=pmids)
```

Comment:
I'm unsure why there is this distinction? Couldn't `result = pd.DataFrame(np.array(output), index=pmids)` in both cases?

Reply:
if the user provided a path, we return a DataFrame containing the memmap itself. If not, we copy it into an in-memory array and return that -- so the role of `np.array(output)` is to make a copy. does that answer the question?
although that might be simplified a bit  with something like 
```python
if not return_memmap:
    output = np.array(output)
return pd.DataFrame(output, index=pmids), masker
```

Reply:
sorry actually that doesn't work -- pandas copies the data into a new numpy array even when copy=False

Reply:
one way we could separate these two options would be to actually have 2 functions: `coordinates_to_maps` that most users would use does not accept a file path and always returns a (in-memory) DataFrame. it, and users if they want to, calls the low-level `coordinates_to_memmapped_maps` that requires a file path and always returns a memmap